### PR TITLE
Fix deadlock on thread termination 

### PR DESCRIPTION
**Pull request by vlantonov**
- Added overrides of exitThreadMain method in derived delegates instead of base implementation.
- Added warning and assertinon in the ThreadDelegate to avoid the force termination of threads.

Related issue: [APPLINK-27410](https://adc.luxoft.com/jira/browse/APPLINK-27410)

### CODE REVIEW

**Comment by vlantonov** [comment]
Please review: @AByzhynar, @OHerasym, @VProdanov, @wolfylambova22, @pvvasilev, @pestOO

File: src/components/transport_manager/src/transport_adapter/threaded_socket_connection.cc [hunk 1]
```diff
@@ -229,6 +229,8 @@ void ThreadedSocketConnection::SocketConnectionDelegate::threadMain() {
 
 void ThreadedSocketConnection::SocketConnectionDelegate::exitThreadMain() {
   SDL_AUTO_TRACE();
+  connection_->terminate_flag_ = true;
+  connection_->unexpected_disconnect_ = true;
 }
```
Comment (VProdanov) [inline src/components/transport_manager/src/transport_adapter/threaded_socket_connection.cc]: @vlantonov Can you please change the order of terminate_flag_ and unexpected_disconnect in void ThreadedSocketConnection::Abort() to match exitThreadMain()

File: src/components/hmi_message_handler/src/mqueue_adapter.cc [hunk 2]
```diff
@@ -70,8 +72,13 @@ class ReceiverThreadDelegate : public threads::ThreadDelegate {
     }
   }
 
+  void exitThreadMain() OVERRIDE {
+    shutdown_requested_ = true;
```
Comment (pestOO) [inline src/components/hmi_message_handler/src/mqueue_adapter.cc]: @vlantonov,
How we stop thread if we are waiting something with `mq_receive`?

Without `pthread_cancel` we need to somehow cancel mq reading.

I see 2 possible solutions:
* Using mqueue with `poll` - it's working on Linux, possible works on QNX.
 * See [TCP example](https://github.com/smartdevicelink/sdl_core/blob/c0107a52570b4b189dcc8aa5a055a1f38ae91c81/src/components/transport_manager/src/transport_adapter/threaded_socket_connection.cc#L209)
*  Using `O_NONBLOCK ` reading and checking result flag

File: src/components/transport_manager/src/bluetooth/bluetooth_device_scanner_posix.cc [hunk 3]
```diff
@@ -497,5 +496,11 @@ void BluetoothDeviceScanner::BluetoothDeviceScannerDelegate::threadMain() {
   scanner_->Thread();
 }
 
+void BluetoothDeviceScanner::BluetoothDeviceScannerDelegate::exitThreadMain() {
+  scanner_->shutdown_requested_ = true;
```
Comment (pestOO) [inline src/components/transport_manager/src/bluetooth/bluetooth_device_scanner_posix.cc]: @vlantonov,
Do we guaranty that all reading in `threadmain`is blocking in less then a few second and non-blocking?
I see a risk the same behavior like in `mqueue_adapter.cc`.

**Comment by vlantonov** [comment]
Please review: @anosach-luxoft

File: src/components/transport_manager/src/bluetooth/bluetooth_device_scanner_win.cc [hunk 4]
```diff
@@ -451,5 +452,11 @@ void BluetoothDeviceScanner::BluetoothDeviceScannerDelegate::threadMain() {
   scanner_->Thread();
 }
 
+void BluetoothDeviceScanner::BluetoothDeviceScannerDelegate::exitThreadMain() {
+  scanner_->shutdown_requested_ = true;
```
Comment (pestOO) [inline src/components/transport_manager/src/bluetooth/bluetooth_device_scanner_win.cc]: @vlantonov,
The same question as above.

File: src/components/transport_manager/src/usb/libusb/usb_handler.cc [hunk 5]
```diff
@@ -320,7 +324,10 @@ void UsbHandler::Thread() {
   SDL_TRACE("enter");
   int completed = 0;
   while (!shutdown_requested_) {
-    libusb_handle_events_completed(libusb_context_, &completed);
+    // all zero timeval struct for non-blocking mode
+    struct timeval max_wait_time = {0, 0};
+    libusb_handle_events_timeout_completed(
```
Comment (pestOO) [inline src/components/transport_manager/src/usb/libusb/usb_handler.cc]: @vlantonov.
Looks like we start using use all CPU resources for this kind of pulling.
Looks like `liusb` also recommend to use `poll` - http://libusb.sourceforge.net/api-1.0/mtasync.html

File: src/components/transport_manager/src/transport_adapter/threaded_socket_connection.cc (hunk 1 above)
Comment (vlantonov) [inline src/components/transport_manager/src/transport_adapter/threaded_socket_connection.cc]: @VProdanov Fixed, see 62c5d87c706c91361468adb794ff7043017a5685

File: src/components/transport_manager/src/bluetooth/bluetooth_device_scanner_posix.cc (hunk 3 above)
Comment (vlantonov) [inline src/components/transport_manager/src/bluetooth/bluetooth_device_scanner_posix.cc]: @pestOO According to the description of the used BlueZ functions - it is guaranteed.

File: src/components/transport_manager/src/bluetooth/bluetooth_device_scanner_win.cc (hunk 4 above)
Comment (vlantonov) [inline src/components/transport_manager/src/bluetooth/bluetooth_device_scanner_win.cc]: @pestOO According to the documentation of Microsoft Bluetooth - it is guaranteed.

File: src/components/transport_manager/src/usb/libusb/usb_handler.cc (hunk 5 above)
Comment (vlantonov) [inline src/components/transport_manager/src/usb/libusb/usb_handler.cc]: @pestOO See 8e59950a3a3aa8cc31fc5731eb3e2fc72d6421a1 According to [Using an event handling thread](http://libusb.sourceforge.net/api-1.0/group__asyncio.html) the blocking `  libusb_handle_events` is unlocked by the call to `libusb_hotplug_deregister_callback`

File: src/components/hmi_message_handler/src/mqueue_adapter.cc (hunk 2 above)
Comment (vlantonov) [inline src/components/hmi_message_handler/src/mqueue_adapter.cc]: @pestOO According to [Advanced UNIX Programming](https://books.google.bg/books?id=vlM72G-1VDkC&dq=mq_receive+poll&hl=bg&source=gbs_navlinks_s) the `poll` will not work - [There's no way to tie a message-waiting system call directly to a file descriptor](https://books.google.bg/books?id=vlM72G-1VDkC&pg=PT421&lpg=PT421&dq=mq_receive+poll&source=bl&ots=kJi7iZk4Hs&sig=PuovRw0DN9Cn_o5N5ukFzXz6O_E&hl=bg&sa=X&ved=0ahUKEwj0pMn97JPPAhUISJoKHaEaDbAQ6AEIWzAH#v=onepage&q=mq_receive%20poll&f=false) So the second solution using `O_NONBLOCK` should be used.

Comment (vlantonov) [inline src/components/hmi_message_handler/src/mqueue_adapter.cc]: @pestOO Fixed, see 9b1e851350176ae8bf460d6467fab2f40cf8804b Please note that `MqueueAdapter` is not used anywhere and related Unit tests are disabled with comment to be fixed in a separate task from @Kozoriz

**Comment by vlantonov** [comment]
@pestOO Commits split, see 44a43aa803a1fc46b271834317023c2c915fba41

**Comment by pestOO** [comment]
@vlantonov, do we need to post Applink jira ticket in each commit?

**Comment by vlantonov** [comment]
@pestOO I cannot remember if there is a specific instruction to put JIRA ticket on each commit, but I do remember that before merging the pull request the commit should be only one and should have JIRA ticket in it. As the review is not completed there may appear additional commits which are to be merged in the one final commit. There will be the JIRA ticket. I split the commits after your request in order review to be more convenient.

File: src/components/transport_manager/test/CMakeLists.txt [hunk 6]
```diff
@@ -59,3 +59,4 @@ set(LIBRARIES
 create_test(transport_manager_test "${SOURCES}" "${LIBRARIES}")
 
 file(COPY smartDeviceLink_test.ini DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
+file(COPY app_info_storage2 DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
```
Comment (vlantonov) [inline src/components/transport_manager/test/CMakeLists.txt]: @AByzhynar Actually it is added. Please take a look at the complete changed files.

Comment (AByzhynar) [inline src/components/transport_manager/test/CMakeLists.txt]: @vlantonov Thank you

File: src/components/transport_manager/src/usb/libusb/usb_handler.cc [hunk 7]
```diff
@@ -90,20 +90,24 @@ UsbHandler::UsbHandler()
...
     libusb_hotplug_deregister_callback(libusb_context_,
                                        arrived_callback_handle_);
     libusb_hotplug_deregister_callback(libusb_context_, left_callback_handle_);
-  }
-  thread_->stop();
-  SDL_INFO("UsbHandler thread finished");
-  if (libusb_context_) {
+
+// TODO(VlAntonov) APPLINK-27938
+// This fix prevents crash under Windows but causes memory leaks
```
Comment (vlantonov) [inline src/components/transport_manager/src/usb/libusb/usb_handler.cc]: @anosach-luxoft Inside the `libusbx` there is an access to invalid data. See details in [APPLINK-27938](https://adc.luxoft.com/jira/browse/APPLINK-27938)

File: src/components/transport_manager/src/usb/libusb/usb_handler.cc [hunk 8]
```diff
@@ -501,5 +505,15 @@ void UsbHandler::UsbHandlerDelegate::threadMain() {
   handler_->Thread();
 }
 
+void UsbHandler::UsbHandlerDelegate::exitThreadMain() {
+  handler_->shutdown_requested_ = true;
```
Comment (vlantonov) [inline src/components/transport_manager/src/usb/libusb/usb_handler.cc]: @anosach-luxoft This thread delegate is initialized as `new UsbHandlerDelegate(this)` in the constructor of `UsbHandler` which supplies valid pointer.

File: src/components/hmi_message_handler/src/mqueue_adapter.cc [hunk 9]
```diff
@@ -48,14 +49,22 @@ class ReceiverThreadDelegate : public threads::ThreadDelegate {
...
-          mq_receive(mqueue_descriptor_, buffer, kMqueueMessageSize, NULL);
+
+      struct timespec current_time;
+      clock_gettime(CLOCK_REALTIME, &current_time);
+
+      const time_t wait_time_sec = 1;
+      current_time.tv_sec += wait_time_sec;
+
+      const ssize_t size = mq_timedreceive(
+          mqueue_descriptor_, buffer, kMqueueMessageSize, NULL, &current_time);
```
Comment (vlantonov) [inline src/components/hmi_message_handler/src/mqueue_adapter.cc]: @anosach-luxoft When I applied similar approach in the `UsbHandlerDelegate` I was warned by @pestOO that this consumes too much CPU time as it checks the queue without pauses. So I preferred timed wait.

**Pull request merged**